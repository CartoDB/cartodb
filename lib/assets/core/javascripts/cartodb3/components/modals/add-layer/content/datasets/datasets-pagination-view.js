var CoreView = require('backbone/core-view');
var PaginationModel = require('../../../../pagination/pagination-model');
var PaginationView = require('../../../../pagination/pagination-view');

/**
 * Responsible for the datasets paginator
 *  ___________________________________________________________________________
 * |                                                                           |
 * |                                             Page 2 of 42 [1] 2 [3][4][5]  |
 * |___________________________________________________________________________|
 *
 */

module.exports = CoreView.extend({
  className: 'DatasetsPaginator',

  initialize: function (opts) {
    if (!opts.routerModel) throw new TypeError('routerModel is required');
    if (!opts.tablesCollection) throw new TypeError('tablesCollection is required');

    this._routerModel = opts.routerModel;
    this._tablesCollection = opts.tablesCollection;

    this.model = new PaginationModel({
      current_page: this._routerModel.get('page')
    });

    this._initBinds();
    this._initViews();
  },

  render: function () {
    this.clearSubViews();
    this.$el.append(this.paginationView.render().el);
    return this;
  },

  _initBinds: function () {
    this.model.bind('change', this.render, this);
    this.model.bind('change:current_page', function () {
      this._routerModel.set('page', this.model.get('current_page'));
    }, this);
    this._tablesCollection.bind('sync', this._updatePaginationModelByCollection, this);
    this._routerModel.bind('change:page', this._updatePaginationModelByRouterModel, this);

    this.add_related_model(this._routerModel);
    this.add_related_model(this._tablesCollection);
    this.add_related_model(this.model);
  },

  _initViews: function () {
    this.paginationView = new PaginationView({
      model: this.model
    });
    this.addView(this.paginationView);
  },

  _updatePaginationModelByCollection: function () {
    this.model.set({
      per_page: this._tablesCollection.getDefaultParam('per_page'),
      total_count: this._tablesCollection.getTotalStat('total_entries')
    });
  },

  _updatePaginationModelByRouterModel: function () {
    this.model.set('current_page', this._routerModel.get('page'));
  }

});
