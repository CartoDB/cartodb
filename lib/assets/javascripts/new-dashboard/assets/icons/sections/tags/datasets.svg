<svg width="14" height="16" xmlns="http://www.w3.org/2000/svg">
  <path d="M7 .5C3.645.5.25 1.53.25 3.5v9c0 1.97 3.395 3 6.75 3s6.75-1.03 6.75-3v-9c0-1.97-3.395-3-6.75-3zM7 2c3.41 0 5.25 1.065 5.25 1.5C12.25 3.935 10.41 5 7 5S1.75 3.935 1.75 3.5C1.75 3.065 3.59 2 7 2zm0 12c-3.41 0-5.25-1.065-5.25-1.5v-1.05c1.302.692 3.28 1.05 5.25 1.05 1.97 0 3.948-.359 5.25-1.05v1.05c0 .435-1.84 1.5-5.25 1.5zm0-3c-3.41 0-5.25-1.065-5.25-1.5V8.45C3.052 9.142 5.03 9.5 7 9.5c1.97 0 3.948-.359 5.25-1.05V9.5c0 .435-1.84 1.5-5.25 1.5zm0-3C3.59 8 1.75 6.935 1.75 6.5V5.45C3.052 6.141 5.03 6.5 7 6.5c1.97 0 3.948-.359 5.25-1.05V6.5C12.25 6.935 10.41 8 7 8z" fill="#2E3C43" fill-rule="nonzero"/>
</svg>
