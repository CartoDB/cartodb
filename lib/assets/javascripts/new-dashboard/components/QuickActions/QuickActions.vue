<template>
  <div class="quick-actions">
    <a href="javascript:void(0)" class="quick-actions-select" @click="toggleDropdown" :class="{'is-active': isOpen, 'has-shadow': hasShadow}">
      <img svg-inline src="new-dashboard/assets/icons/common/options.svg">
    </a>
    <div class="quick-actions-dropdown" :class="{'is-active' : isOpen}" v-if="isOpen" v-click-outside="closeDropdown" @click="killEvent">
      <h6 class="quick-actions-title text is-semibold is-xsmall is-txtSoftGrey">{{ $t(`QuickActions.title`) }}</h6>
      <ul>
        <li v-for="action in actions" :key="action.name" v-if="!action.shouldBeHidden">
          <a href="#" class="action text is-caption" :class="{'is-txtPrimary': !action.isDestructive, 'is-txtAlert': action.isDestructive}" @click="emitEvent(action.event)">{{action.name}}</a>
        </li>
      </ul>
    </div>
  </div>
</template>

<script>

export default {
  name: 'QuickActions',
  data: function () {
    return {
      isOpen: false
    };
  },
  props: {
    actions: Array,
    hasShadow: {
      type: Boolean,
      default: true
    }
  },
  methods: {
    emitEvent (action) {
      this.$emit(action);
    },
    toggleDropdown () {
      this.isOpen = !this.isOpen;
      if (this.isOpen) {
        this.$emit('open');
      } else {
        this.$emit('close');
      }
    },
    closeDropdown () {
      this.isOpen = false;
      this.$emit('close');
    },
    killEvent (event) {
      event.preventDefault();
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.quick-actions-select {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  background: $white;

  &.is-active {
    background-color: $primary-color;

    .path {
      fill: $white;
    }
  }

  &.has-shadow {
    box-shadow: 1px 1px 1px rgba(0, 0, 0, 0.12);
  }

  &:hover {
    border: 1px solid $primary-color;
  }
}

.quick-actions-dropdown {
  position: absolute;
  z-index: 2;
  right: 0;
  width: 260px;
  margin-top: 8px;
  border: 1px solid $grey;
  border-radius: 2px;
  background-color: $white;
  cursor: default;
}

.quick-actions-title {
  margin-top: 16px;
  margin-left: 24px;
  text-transform: uppercase;
}

.action {
  display: block;
  padding: 14px 24px 16px;
  border-bottom: 1px solid $softblue;
}
</style>
