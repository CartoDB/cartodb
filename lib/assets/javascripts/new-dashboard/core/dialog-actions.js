// Dialogs
import Dialog from 'new-dashboard/components/Backbone/Dialog.vue';
import CreateDialog from 'new-dashboard/components/Backbone/Dialogs/CreateDialog';
import ChangePrivacy from 'new-dashboard/components/Backbone/Dialogs/ChangePrivacy';
import DeleteDialog from 'new-dashboard/components/Backbone/Dialogs/DeleteDialog';
import DuplicateMap from 'new-dashboard/components/Backbone/Dialogs/DuplicateMap';
import LockVisualization from 'new-dashboard/components/Backbone/Dialogs/LockVisualization';
import MapMetadata from 'new-dashboard/components/Backbone/Dialogs/MapMetadata';
import DatasetMetadata from 'new-dashboard/components/Backbone/Dialogs/DatasetMetadata';
import Share from 'new-dashboard/components/Backbone/Dialogs/Share';

export function editMapMetadata (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <MapMetadata
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, MapMetadata },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function editDatasetMetadata (dataset, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <DatasetMetadata
          :dataset="dataset"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['dataset'],
      components: { Dialog, DatasetMetadata },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { dataset }
  );
}

export function createMap (selectedDatasets, backgroundPollingView, mamufasImportView) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <CreateDialog :selectedItems="selectedItems" :backgroundPollingView="backgroundPollingView" :mamufasImportView="mamufasImportView" @close="$emit('close')"/>
      </Dialog>`,
      props: ['selectedItems', 'backgroundPollingView', 'mamufasImportView'],
      components: { Dialog, CreateDialog }
    },
    {
      selectedItems: selectedDatasets,
      backgroundPollingView,
      mamufasImportView
    }
  );
}

export function changePrivacy (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <ChangePrivacy
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, ChangePrivacy },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function shareVisualization (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <Share
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, Share },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function duplicateVisualization (visualization) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DuplicateMap :visualization="visualization" @close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, DuplicateMap }
    },
    { visualization }
  );
}

export function changeLockState (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockVisualization
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, LockVisualization },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function changeVisualizationsLockState (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockVisualization
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, LockVisualization },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function deleteVisualization (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function deleteVisualizations (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}
