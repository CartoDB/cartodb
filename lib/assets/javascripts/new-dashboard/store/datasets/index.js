import CartoNode from 'carto-node';
import toObject from 'new-dashboard/utils/to-object';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/filters';

const client = new CartoNode.AuthenticatedClient();

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'updated_at',
  orderDirection: 'desc'
};

const datasets = {
  namespaced: true,
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 1
  },
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setDatasets (state, datasets) {
      state.list = toObject(datasets.visualizations, 'id');
      state.metadata = {
        total_entries: datasets.total_entries,
        total_likes: datasets.total_likes,
        total_shared: datasets.total_shared,
        total_locked: datasets.total_locked,
        total_user_entries: datasets.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setDatasetAttributes (state, {datasetId, datasetAttributes}) {
      Object.assign(state.list[datasetId], datasetAttributes);
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setOrder (state, orderOptions) {
      state.order = orderOptions.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOptions.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage) {
      state.resultsPerPage = resultsPerPage;
    },

    updateLike (state, {datasetId, liked}) {
      state.list[datasetId].liked = liked;
    },

    updateNumberLikes (state, {datasetId, likes}) {
      state.list[datasetId].likes = likes;
    }
  },
  actions: {
    fetchDatasets (context) {
      context.commit('setFetchingState');

      const params = {
        ...Filters[context.state.filterType],
        types: 'table',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage,
        with_dependent_visualizations: 10
      };

      client.getVisualization('',
        params,
        function (err, _, data) {
          if (err) {
            context.commit('setRequestError', err);
            return;
          }
          context.commit('setDatasets', data);
          context.commit('setPagination', context.state.page);
        });
    },

    filterDatasets (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetchDatasets');
    },

    orderDatasets (context, orderOptions) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOptions);
      context.dispatch('fetchDatasets');
    },

    like (context, dataset) {
      const currentLikeStatus = dataset.liked;
      context.commit('updateLike', { datasetId: dataset.id, liked: true });
      client.like(dataset.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { datasetId: dataset.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { datasetId: dataset.id, likes: data.likes });
        }
      );
    },

    deleteLike (context, dataset) {
      const currentLikeStatus = dataset.liked;
      context.commit('updateLike', { datasetId: dataset.id, liked: false });
      client.deleteLike(dataset.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { datasetId: dataset.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { datasetId: dataset.id, likes: data.likes });
        }
      );
    },

    updateDataset (context, datasetOptions) {
      context.commit('setDatasetAttributes', datasetOptions);
    },

    setResultsPerPage (context, perPage) {
      context.commit('setResultsPerPage', perPage);
    },

    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', { order: options.order, direction: options.order_direction });
      context.dispatch('fetchDatasets');
    },

    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', { order: DEFAULT_VALUES.order, direction: DEFAULT_VALUES.orderDirection });
    }
  }
};

export default datasets;
