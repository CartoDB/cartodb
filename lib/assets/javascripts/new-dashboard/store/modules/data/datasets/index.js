import Filters from 'new-dashboard/core/configuration/filters';

import * as VisualizationActions from '../../../actions/visualizations';
import * as VisualizationMutations from '../../../mutations/visualizations';

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'favorited,updated_at',
  orderDirection: 'desc,desc'
};

const datasets = {
  namespaced: true,
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 1
  },
  mutations: {
    setFetchingState: VisualizationMutations.setFetchingState,
    setFilterType: VisualizationMutations.setFilterType,
    setOrder: VisualizationMutations.setOrder(DEFAULT_VALUES),
    setPagination: VisualizationMutations.setPagination,
    setResultsPerPage: VisualizationMutations.setResultsPerPage,
    setRequestError: VisualizationMutations.setRequestError,
    setVisualizations: VisualizationMutations.setVisualizations,
    updateNumberLikes: VisualizationMutations.updateNumberLikes,
    updateVisualization: VisualizationMutations.updateVisualization
  },
  actions: {
    deleteLike: VisualizationActions.deleteLike,
    filter: VisualizationActions.filter,
    like: VisualizationActions.like,
    order: VisualizationActions.order,
    resetFilters: VisualizationActions.resetFilters(DEFAULT_VALUES),
    setResultsPerPage: VisualizationActions.setResultsPerPage,
    setURLOptions: VisualizationActions.setURLOptions,
    updateVisualization: VisualizationActions.updateVisualization,

    fetch (context) {
      const params = {
        ...Filters[context.state.filterType],
        types: 'table',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage,
        with_dependent_visualizations: 10
      };

      VisualizationActions.fetchVisualizations(context, params);
    }
  }
};

export default datasets;
