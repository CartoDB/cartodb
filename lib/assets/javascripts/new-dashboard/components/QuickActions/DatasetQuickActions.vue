<template>
  <QuickActions
    :actions="actions[actionMode]"
    :hasShadow="false"
    v-on="getEventListeners()"
    ref="quickActions"
    @open="openQuickactions"
    @close="closeQuickactions" />
</template>

<script>
import * as Table from 'new-dashboard/core/table';
import QuickActions from 'new-dashboard/components/QuickActions/QuickActions';
import * as DialogActions from 'new-dashboard/core/dialog-actions';

export default {
  name: 'DatasetQuickActions',
  inject: ['backboneViews'],
  components: {
    QuickActions
  },
  props: {
    dataset: Object,
    isShared: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    actions () {
      return {
        mine: [
          { name: this.$t('QuickActions.createMap'), event: 'createMap' },
          { name: this.$t('QuickActions.duplicate'), event: 'duplicateDataset' },
          { name: this.$t('QuickActions.share'), event: 'shareVisualization', shouldBeHidden: !this.isUserInsideOrganization },
          { name: this.$t('QuickActions.editInfo'), event: 'editInfo' },
          { name: this.$t('QuickActions.manageTags'), event: 'manageTags' },
          { name: this.$t('QuickActions.changePrivacy'), event: 'changePrivacy' },
          { name: this.$t('QuickActions.lock'), event: 'lockDataset' },
          { name: this.$t('QuickActions.delete'), event: 'deleteDataset', isDestructive: true }
        ],
        shared: [
          { name: this.$t('QuickActions.createMap'), event: 'createMap' },
          { name: this.$t('QuickActions.duplicate'), event: 'duplicateDataset' }
        ],
        locked: [
          { name: this.$t('QuickActions.duplicate'), event: 'duplicateDataset' },
          { name: this.$t('QuickActions.unlock'), event: 'unlockDataset' }
        ]
      };
    },
    actionMode () {
      if (this.dataset.locked) {
        return 'locked';
      }
      if (this.isShared) {
        return 'shared';
      }
      return 'mine';
    },
    isUserInsideOrganization () {
      const userOrganization = this.$store.state.user.organization;
      return userOrganization && userOrganization.id;
    }
  },
  methods: {
    getActionHandlers () {
      return {
        deselectAll: () => {},
        fetchList: () => {
          this.$store.dispatch('datasets/fetchDatasets');
        },
        updateVisualization: (model) => {
          this.$store.dispatch('datasets/updateDataset', { datasetId: model.get('id'), datasetAttributes: model.attributes });
        }
      };
    },
    getEventListeners () {
      const events = this.actions[this.actionMode].map(action => action.event);

      return events.reduce(
        (eventListeners, action) => {
          eventListeners[action] = this[action].bind(this);
          return eventListeners;
        }, {}
      );
    },
    showModal (componentDefinition, componentPropsData) {
      this.$modal.show(
        componentDefinition,
        componentPropsData,
        { width: '100%', height: '100%' }
      );
    },
    openQuickactions () {
      this.$emit('open');
    },
    closeQuickactions () {
      this.$emit('close');
    },
    closeDropdown () {
      this.$refs.quickActions.closeDropdown();
    },
    editInfo () {
      DialogActions.editDatasetMetadata.apply(this, [this.dataset, this.getActionHandlers()]);
      this.closeDropdown();
    },
    createMap () {
      DialogActions.createMap.apply(this, [
        [this.dataset],
        this.backboneViews.backgroundPollingView.getBackgroundPollingView(),
        this.backboneViews.mamufasImportView.getView()
      ]);
    },
    changePrivacy () {
      DialogActions.changePrivacy.apply(this, [this.dataset, this.getActionHandlers()]);
      this.closeDropdown();
    },
    manageTags () {
      DialogActions.editDatasetMetadata.apply(this, [this.dataset, this.getActionHandlers()]);
      this.closeDropdown();
    },
    duplicateDataset () {
      const bgPollingView = this.backboneViews.backgroundPollingView.getBackgroundPollingView();

      bgPollingView._addDataset({
        type: 'duplication',
        table_name: `${Table.getUnqualifiedName(this.dataset.name)}_copy`,
        value: this.dataset.name,
        create_vis: false
      });
      this.closeDropdown();
    },
    unlockDataset () {
      DialogActions.changeLockState.apply(this, [this.dataset, 'datasets', this.getActionHandlers()]);
      this.closeDropdown();
    },
    lockDataset () {
      DialogActions.changeLockState.apply(this, [this.dataset, 'datasets', this.getActionHandlers()]);
      this.closeDropdown();
    },
    deleteDataset () {
      DialogActions.deleteVisualization.apply(this, [this.dataset, 'datasets', this.getActionHandlers()]);
      this.closeDropdown();
    },
    shareVisualization () {
      DialogActions.shareVisualization.apply(this, [this.dataset, this.getActionHandlers()]);
      this.closeDropdown();
    }
  }
};
</script>
