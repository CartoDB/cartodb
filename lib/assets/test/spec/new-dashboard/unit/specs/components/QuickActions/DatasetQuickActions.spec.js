import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import DatasetQuickActions from 'new-dashboard/components/QuickActions/DatasetQuickActions';
import QuickActions from 'new-dashboard/components/QuickActions/QuickActions';
import datasets from '../../fixtures/datasets';
import datasetsLocked from '../../fixtures/datasetsLocked';

import * as DialogActions from '../__mocks__/dialog-actions';
import usersArray from '../../fixtures/users';

// Backbone Models
import ConfigModel from 'dashboard/data/config-model';
import UserModel from 'dashboard/data/user-model';

const localVue = createLocalVue();
localVue.use(Vuex);

const $t = key => key;

function configCartoModels (attributes = {}) {
  const user = new UserModel(attributes.user || usersArray[0]);
  const config = new ConfigModel(attributes.config || { maps_api_template: 'http://{user}.example.com' });
  return { user, config };
}

let $cartoModels = configCartoModels();

let store = new Vuex.Store({
  state: {
    user: {
      organization: { id: 'fake-organization-id' }
    }
  }
});

const addDatasetSpy = jest.fn();
const backgroundPollingView = {
  getBackgroundPollingView () {
    return { _addDataset: addDatasetSpy };
  }
};

describe('DatasetQuickAction.vue', () => {
  it('should render open dropdown with all actions for non shared non locked datasets', () => {
    const dataset = datasets.visualizations[0];
    const datasetQuickAction = createDatasetQuickActionComponent(dataset);
    const quickActionDropdown = datasetQuickAction.find(QuickActions);
    quickActionDropdown.setData({ isOpen: true });

    expect(quickActionDropdown).toMatchSnapshot();
  });

  it('should render open dropdown with actions for locked datasets', () => {
    const dataset = datasetsLocked.visualizations[0];
    const datasetQuickAction = createDatasetQuickActionComponent(dataset);
    const quickActionDropdown = datasetQuickAction.find(QuickActions);
    quickActionDropdown.setData({ isOpen: true });

    expect(quickActionDropdown).toMatchSnapshot();
  });

  describe('computed', () => {
    it('should return "locked" action name when a map is locked', () => {
      const dataset = datasetsLocked.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      const result = datasetQuickAction.vm.actionMode;

      expect(result).toBe('locked');
    });

    it('should return "shared" action name when a map is shared', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.setProps({ isShared: true });
      const result = datasetQuickAction.vm.actionMode;

      expect(result).toBe('shared');
    });

    it('should return "mine" action name when a map is not locked or shared', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      const result = datasetQuickAction.vm.actionMode;

      expect(result).toBe('mine');
    });
  });

  describe('methods', () => {
    it('should call the function to show modal', () => {
      const dataset = datasets.visualizations[0];
      const $modal = { show: jest.fn() };
      const datasetQuickAction = createDatasetQuickActionComponent(dataset, { $modal });

      datasetQuickAction.vm.showModal();

      expect($modal.show).toHaveBeenCalled();
    });

    it('should emit event open when calling openQuickactions method', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      expect(datasetQuickAction.emitted('open')).toBeFalsy();

      datasetQuickAction.vm.openQuickactions();

      expect(datasetQuickAction.emitted('open').length).toBe(1);
    });

    it('should emit event close when calling closeQuickactions method', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      expect(datasetQuickAction.emitted('close')).toBeFalsy();

      datasetQuickAction.vm.closeQuickactions();

      expect(datasetQuickAction.emitted('close').length).toBe(1);
    });

    it('should open edit metadata modal', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.editInfo();

      expect(DialogActions.editDatasetMetadata).toHaveBeenCalled();
    });

    it('should open change privacy modal', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.changePrivacy();

      expect(DialogActions.changePrivacy).toHaveBeenCalled();
    });

    it('should open manage tags modal', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.manageTags();

      expect(DialogActions.editDatasetMetadata).toHaveBeenCalled();
    });

    it('should launch share with colleagues action', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.shareVisualization();

      expect(DialogActions.shareVisualization).toHaveBeenCalled();
    });

    it('should launch duplicate dataset action', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.duplicateDataset();

      expect(addDatasetSpy).toHaveBeenCalled();
    });

    it('should launch unlock dataset action', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.unlockDataset();

      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should launch lock dataset action', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.lockDataset();

      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should launch delete dataset action', () => {
      const dataset = datasets.visualizations[0];
      const datasetQuickAction = createDatasetQuickActionComponent(dataset);
      datasetQuickAction.vm.closeDropdown = jest.fn();

      datasetQuickAction.vm.deleteDataset();

      expect(DialogActions.deleteVisualization).toHaveBeenCalled();
    });
  });
});

function createDatasetQuickActionComponent (dataset, additionalMocks) {
  return mount(DatasetQuickActions, {
    store,
    localVue,
    propsData: { dataset, hasShadow: false },
    mocks: {
      $cartoModels,
      $t,
      ...additionalMocks,
      backboneViews: {
        backgroundPollingView
      }
    }
  });
}
