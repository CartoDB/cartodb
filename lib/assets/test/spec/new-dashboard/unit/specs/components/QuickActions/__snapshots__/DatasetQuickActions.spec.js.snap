// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetQuickAction.vue should render open dropdown with actions for locked datasets 1`] = `
<div class="quick-actions"><a href="javascript:void(0)" class="quick-actions-select is-active"><img svg-inline="" src="new-dashboard/assets/icons/common/options.svg"></a>
  <div class="quick-actions-dropdown is-active">
    <h6 class="quick-actions-title text is-semibold is-xsmall is-txtSoftGrey">QuickActions.title</h6>
    <ul>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.duplicate</a></li>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.unlock</a></li>
    </ul>
  </div>
</div>
`;

exports[`DatasetQuickAction.vue should render open dropdown with all actions for non shared non locked datasets 1`] = `
<div class="quick-actions"><a href="javascript:void(0)" class="quick-actions-select is-active"><img svg-inline="" src="new-dashboard/assets/icons/common/options.svg"></a>
  <div class="quick-actions-dropdown is-active">
    <h6 class="quick-actions-title text is-semibold is-xsmall is-txtSoftGrey">QuickActions.title</h6>
    <ul>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.createMap</a></li>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.duplicate</a></li>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.share</a></li>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.editInfo</a></li>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.manageTags</a></li>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.changePrivacy</a></li>
      <li><a href="#" class="action text is-caption is-txtPrimary">QuickActions.lock</a></li>
      <li><a href="#" class="action text is-caption is-txtAlert">QuickActions.delete</a></li>
    </ul>
  </div>
</div>
`;
